### ISSUE TITLE
Replace `TSMappedType::type_parameter` with dedicated fields to match TSESTree

### ISSUE THREAD

**Issue by Dunqing** [oxc-project/oxc#15101]
## Summary

Currently, `TSMappedType` uses `TSTypeParameter` to represent its type parameter, which includes several fields that are syntactically invalid in mapped types (`r#const`, `default`). We should replace `type_parameter` with dedicated fields that accurately represent the allowed syntax and match the TSESTree specification.

## Current Implementation

```rust
pub struct TSMappedType<'a> {
    pub span: Span,
    pub type_parameter: Box<'a, TSTypeParameter<'a>>,
    pub name_type: Option<TSType<'a>>,
    pub type_annotation: Option<TSType<'a>>,
    pub optional: Option<TSMappedTypeModifierOperator>,
    pub readonly: Option<TSMappedTypeModifierOperator>,
    pub scope_id: Cell<Option<ScopeId>>,
}
```

## Problem

1. **Semantic Mismatch**: `TSTypeParameter` includes fields that are not valid in mapped type syntax:
   - `r#const` - mapped types cannot use `const` modifier
   - `default` - mapped types cannot have default types
   - `in` - variance annotation not applicable to mapped types
   - `out` - variance annotation not applicable to mapped types

2. **TSESTree Incompatibility**: The current structure doesn't match the TSESTree specification, which defines mapped types with separate `key` and `constraint` fields rather than a full type parameter.

This creates both a semantic mismatch where the AST structure allows syntactically invalid constructs and a compatibility issue with TypeScript tooling.

## Proposed Solution

Replace `type_parameter: Box<'a, TSTypeParameter<'a>>` with two explicit fields to match TSESTree:

````rust
#[ast(visit)]
#[scope]
#[derive(Debug)]
#[generate_derive(CloneIn, Dummy, TakeIn, GetSpan, GetSpanMut, ContentEq, ESTree)]
pub struct TSMappedType<'a> {
    pub span: Span,
    /// The type parameter name (e.g., `P` in `[P in keyof T]`)
    pub key: BindingIdentifier<'a>,
    /// The constraint (e.g., `keyof T` in `[P in keyof T]`)
    pub constraint: TSType<'a>,
    pub name_type: Option<TSType<'a>>,
    pub type_annotation: Option<TSType<'a>>,
    pub optional: Option<TSMappedTypeModifierOperator>,
    pub readonly: Option<TSMappedTypeModifierOperator>,
    pub scope_id: Cell<Option<ScopeId>>,
}
````

Note: The existing `#[estree(add_fields(key = TSMappedTypeKey, constraint = TSMappedTypeConstraint))]` can be removed as these will now be real fields.

## Benefits

1. **Type Safety**: The AST structure will only allow syntactically valid constructs
2. **TSESTree Compatibility**: Matches the structure expected by TypeScript tooling and TypeScript-ESLint
3. **Clarity**: The intent and constraints are explicit in the type definition
4. **Maintainability**: Reduces confusion about which `TSTypeParameter` fields are applicable
5. **Correctness**: Prevents the possibility of creating invalid AST nodes

## Example

For the mapped type:
```ts
type Readonly<T> = { readonly [P in keyof T]: T[P] }
```

- `key` would be the identifier `P` (as `BindingIdentifier`)
- `constraint` would be the type `keyof T` (as `TSType`)

## TSESTree Compatibility

This change aligns Oxc's AST with [typescript-eslint's TSESTree specification](https://github.com/typescript-eslint/typescript-eslint/blob/55ca033ee88cc95cfac4ad6dea2257fbeb1d4657/packages/ast-spec/src/type/TSMappedType/spec.ts), which defines `TSMappedType` with `key` and `constraint` fields. This could help us to remove specific converting logic from OXC AST to ESTree

**Comment by overlookmotel** [comment]
> This could help us to remove specific converting logic from OXC AST to ESTree

Yes please!
